Delete node and client ALCs on --purge option.

lbakerchef opened the pull request:
Node and client ACLs were not being sync'd on --purge option (the option appeared to only work with user ACLs).  This fixes the issue.

Thanks to @marcparadise for consulting on the issue.

For some context as to why this might be broken:
https://github.com/chef/knife-ec-backup/pull/94

## Related Issues
https://github.com/chef/chef/pull/12458

## Acceptance Criteria
- [x] Node and client ACLs properly purged on --purge option.
- [x] Test if someone on 12.x (12.17.15) can backup with gem that ships with 12.17.15 and restore on later version (14x) using hand-patched gem that comes with that later version (14x).

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

lbakerchef commented:
Test depicting working cleanup and sync of nodes, clients, and users, with proper cleanup of ACLs.
```
root@ip-10-0-4-202:~# knife node create node1
Created node[node1]
root@ip-10-0-4-202:~# knife node create node2
Created node[node2]
root@ip-10-0-4-202:~# knife client create client1
Created client[client1]
root@ip-10-0-4-202:~# knife client create client2
Created client[client2]
root@ip-10-0-4-202:~# knife opc user create user1 user 1 user1@example.com abc123
root@ip-10-0-4-202:~# knife opc user create user2 user 2 user2@example.com abc123
root@ip-10-0-4-202:~# knife node list
node1
node2
root@ip-10-0-4-202:~# knife client list
4thcoffee-validator
client1
client2
root@ip-10-0-4-202:~# chef-server-ctl user-list
janedoe
pivotal
user1
user2
root@ip-10-0-4-202:~# knife ec backup /backup --sql-db automate-cs-oc-erchef --purge -V -s https://ip-10-0-4-202.us-west-2.compute.internal
Downloading Users
Downloading organization object for 4thcoffee from https://ip-10-0-4-202.us-west-2.compute.internal/organizations/4thcoffee
Copying /acls/groups/billing-admins.json
Created /acls
Created /acls/groups
Created /acls/groups/billing-admins.json
Copying /groups/billing-admins.json
Created /groups
Created /groups/billing-admins.json
Copying /groups/admins.json
Created /groups/admins.json
Copying /acls/groups/public_key_read_access.json
Created /acls/groups/public_key_read_access.json
Copying /groups/public_key_read_access.json
Created /groups/public_key_read_access.json
Copying /acls
Created /acls/environments
Created /acls/containers
Created /acls/clients
Created /acls/cookbooks
Created /acls/nodes
Created /acls/cookbook_artifacts
Created /acls/data_bags
Created /acls/roles
Created /acls/policies
Created /acls/policy_groups
Created /acls/nodes/node1.json
Created /acls/containers/containers.json
Created /acls/clients/4thcoffee-validator.json
Created /acls/containers/clients.json
Created /acls/containers/cookbook_artifacts.json
Created /acls/organization.json
Created /acls/groups/000000000000ec9aba30060322d83390.json
Created /acls/environments/_default.json
Created /acls/clients/client1.json
Created /acls/nodes/node2.json
Created /acls/containers/cookbooks.json
Created /acls/containers/data.json
Created /acls/groups/admins.json
Created /acls/containers/environments.json
Created /acls/clients/client2.json
Created /acls/containers/groups.json
Created /acls/containers/policies.json
Created /acls/containers/nodes.json
Created /acls/containers/policy_groups.json
Created /acls/containers/roles.json
Created /acls/containers/sandboxes.json
Created /acls/groups/clients.json
Created /acls/groups/users.json
Copying /clients
Created /clients
Created /clients/client1.json
Created /clients/client2.json
Created /clients/4thcoffee-validator.json
Copying /containers
Created /containers
Created /containers/data.json
Created /containers/cookbook_artifacts.json
Created /containers/cookbooks.json
Created /containers/containers.json
Created /containers/nodes.json
Created /containers/policy_groups.json
Created /containers/environments.json
Created /containers/policies.json
Created /containers/groups.json
Created /containers/clients.json
Created /containers/roles.json
Created /containers/sandboxes.json
Copying /cookbook_artifacts
Created /cookbook_artifacts
Copying /cookbooks
Created /cookbooks
Copying /data_bags
Created /data_bags
Copying /environments
Created /environments
Created /environments/_default.json
Copying /groups
Created /groups/clients.json
Created /groups/000000000000ec9aba30060322d83390.json
Created /groups/users.json
Copying /invitations.json
Created /invitations.json
Copying /members.json
Created /members.json
Copying /nodes
Created /nodes
Created /nodes/node1.json
Created /nodes/node2.json
Copying /org.json
Copying /policies
Created /policies
Copying /policy_groups
Created /policy_groups
Copying /roles
Created /roles
** Finished **
root@ip-10-0-4-202:~# knife node delete node1 -y
Deleted node[node1]
root@ip-10-0-4-202:~# knife node delete node2 -y
Deleted node[node2]
root@ip-10-0-4-202:~# knife client delete client1 -y
Deleted client[client1]
root@ip-10-0-4-202:~# knife client delete client2 -y
root@ip-10-0-4-202:~# chef-server-ctl user-delete user1 -y
Checking organization memberships...
Deleting user user1.
root@ip-10-0-4-202:~# chef-server-ctl user-delete user2 -y
Checking organization memberships...
Deleting user user2.
root@ip-10-0-4-202:~# knife node list

root@ip-10-0-4-202:~# knife client list
4thcoffee-validator
root@ip-10-0-4-202:~# chef-server-ctl user-list
janedoe
pivotal
root@ip-10-0-4-202:~# ls /backup/organizations/4thcoffee/acls/clients
4thcoffee-validator.json  client1.json  client2.json
root@ip-10-0-4-202:~# ls /backup/organizations/4thcoffee/acls/nodes
node1.json  node2.json
root@ip-10-0-4-202:~# ls /backup/user_acls
janedoe.json  pivotal.json  user1.json  user2.json
root@ip-10-0-4-202:~# knife ec backup /backup --sql-db automate-cs-oc-erchef --purge -V -s https://ip-10-0-4-202.us-west-2.compute.internal
Downloading Users
Deleting user user1 from local backup (purge is on)
Deleting user user2 from local backup (purge is on)
Downloading organization object for 4thcoffee from https://ip-10-0-4-202.us-west-2.compute.internal/organizations/4thcoffee
Copying /acls/groups/billing-admins.json
Copying /groups/billing-admins.json
Copying /groups/admins.json
Copying /acls/groups/public_key_read_access.json
Copying /groups/public_key_read_access.json
Copying /acls
Deleted extra entry /acls/clients/client1.json (purge is on)
Deleted extra entry /acls/clients/client2.json (purge is on)
Deleted extra entry /acls/nodes/node1.json (purge is on)
Deleted extra entry /acls/nodes/node2.json (purge is on)
Copying /clients
Deleted extra entry /clients/client2.json (purge is on)
Deleted extra entry /clients/client1.json (purge is on)
Copying /containers
Copying /cookbook_artifacts
Copying /cookbooks
Copying /data_bags
Copying /environments
Copying /groups
Updated /groups/clients.json
Copying /invitations.json
Copying /members.json
Copying /nodes
Deleted extra entry /nodes/node2.json (purge is on)
Deleted extra entry /nodes/node1.json (purge is on)
Copying /org.json
Copying /policies
Copying /policy_groups
Copying /roles
** Finished **
root@ip-10-0-4-202:~# ls /backup/organizations/4thcoffee/acls/nodes
root@ip-10-0-4-202:~# ls /backup/organizations/4thcoffee/acls/clients/
4thcoffee-validator.json
root@ip-10-0-4-202:~# ls /backup/user_acls
janedoe.json  pivotal.json
```

PrajaktaPurohit commented on lib/chef/knife/ec_backup.rb:
```diff
@@ -226,7 +226,7 @@ def download_org_data(name)
           top_level_acls  = chef_fs_paths('/acls/*.json', chef_fs_config, [])
           acl_paths       = chef_fs_paths('/acls/*/*', chef_fs_config, exclude_list)
           group_paths     = chef_fs_paths('/groups/*', chef_fs_config, exclude_list)
```
are these being used anywhere after the change?

marcparadise replied:
The other thing to look at is if this prevents us from excluding billing-admins and public_key_read_access.

lbakerchef replied:
@marcparadise Can you elaborate on excluding billing-admins and public_key_read_access?  Can you point me to something explaining how to do it? (checked docs on `knife ec backup/restore` but didn't see anything).

PrajaktaPurohit replied:
does not look like the top_level_acls, acl_paths and group_paths are being used anywhere. They can be deleted if they are not.

PrajaktaPurohit commented:
https://github.com/chef/chef-server/issues/2842

PrajaktaPurohit commented:
Do we need to update the tests at : https://github.com/chef/knife-ec-backup/blob/main/spec/chef/knife/ec_backup_spec.rb?

lbakerchef commented:
Baseline test, unpatched code, chef-server 12.19.31.
The bug is present on 12.x.
```
# vagrant box / chef-server 12.19.31 / ubuntu1804 / unpatched code / baseline test

mac:dev lbaker$ mixlib-install download chef-server -c stable -a x86_64 -p ubuntu -l 18.04 -v 12.19.31
Starting download https://packages.chef.io/files/stable/chef-server/12.19.31/ubuntu/18.04/chef-server-core_12.19.31-1_amd64.deb
Download saved to /Users/lbaker/chef-server/dev/chef-server-core_12.19.31-1_amd64.deb
mac:dev lbaker$ mv /Users/lbaker/chef-server/dev/chef-server-core_12.19.31-1_amd64.deb dotfiles/
mac:dev lbaker$ VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant destroy && VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant up chef-server
mac:dev lbaker$ vagrant ssh
vagrant@api:~$ sudo -i
root@api:~# dpkg -i /vagrant/dotfiles/chef-server-core_12.19.31-1_amd64.deb
Selecting previously unselected package chef-server-core.
(Reading database ... 39785 files and directories currently installed.)
Preparing to unpack .../chef-server-core_12.19.31-1_amd64.deb ...
Unpacking chef-server-core (12.19.31-1) ...
Setting up chef-server-core (12.19.31-1) ...
root@api:~# chef-server-ctl reconfigure -y 
Chef Client finished, 495/1084 resources updated in 03 minutes 45 seconds
Chef Server Reconfigured!
root@api:~# chef-server-ctl version
12.19.31
root@api:~# alias knife='/opt/opscode/bin/knife'
root@api:~# export EDITOR=vim
root@api:~# mkdir .chef
root@api:~# chef-server-ctl user-create janedoe Jane Doe janed@example.com 'abc123' --filename ~/.chef/janedoe.pem
root@api:~# chef-server-ctl org-create 4thcafe 'Fourth Cafe, Inc.' --association_user janedoe --filename ~/.chef/4thcafe-validator.pem
root@api:~# echo "[default]
> client_name     = 'janedoe'
> client_key      = '~/.chef/janedoe.pem'
> chef_server_url = 'https://api.chef-server.dev/organizations/4thcafe'">~/.chef/credentials
root@api:~# knife ssl fetch
WARNING: Certificates from api.chef-server.dev will be fetched and placed in your trusted_cert
directory (/root/.chef/trusted_certs).

Knife has no means to verify these are the correct certificates. You should
verify the authenticity of these certificates after downloading.

Adding certificate for api_chef-server_dev in /root/.chef/trusted_certs/api_chef-server_dev.crt
root@api:~# knife ssl check
Connecting to host api.chef-server.dev:443
Successfully verified certificates from `api.chef-server.dev'
root@api:~# knife node create node1
Created node[node1]
root@api:~# knife node create node2
Created node[node2]
root@api:~# knife client create client1
Created client[client1]
root@api:~# knife client create client2
Created client[client2]
root@api:~# chef-server-ctl user-create user1 user 1 user1@example.com 'abc123'
root@api:~# chef-server-ctl user-create user2 user 2 user2@example.com 'abc123'
root@api:~# knife node list
node1
node2
root@api:~# knife client list
4thcafe-validator
client1
client2
root@api:~# chef-server-ctl user-list
janedoe
pivotal
user1
user2
root@api:~# knife ec backup /backup --purge -V -s https://api.chef-server.dev
root@api:~# knife node delete node1 -y
Deleted node[node1]
root@api:~# knife node delete node2 -y
Deleted node[node2]
root@api:~# knife client delete client1 -y
Deleted client[client1]
root@api:~# knife client delete client2 -y
Deleted client[client2]
root@api:~# chef-server-ctl user-delete user1 -y
Checking organization memberships...
Deleting user user1.
root@api:~# chef-server-ctl user-delete user2 -y
Checking organization memberships...
Deleting user user2.
root@api:~# knife node list

root@api:~# knife client list
4thcafe-validator
root@api:~# chef-server-ctl user-list
janedoe
pivotal
root@api:~# ls /backup/organizations/4thcafe/acls/clients/
4thcafe-validator.json  client1.json  client2.json
root@api:~# ls /backup/organizations/4thcafe/acls/nodes/
node1.json  node2.json
root@api:~# ls /backup/user_acls
janedoe.json  pivotal.json  user1.json  user2.json
root@api:~# knife ec backup /backup --purge -V -s https://api.chef-server.dev
Downloading Users
Deleting user user1 from local backup (purge is on)
Deleting user user2 from local backup (purge is on)
Downloading organization object for 4thcafe from https://api.chef-server.dev/organizations/4thcafe
Copying /acls/groups/billing-admins.json
Copying /groups/billing-admins.json
Copying /groups/admins.json
Copying /acls/groups/public_key_read_access.json
Copying /groups/public_key_read_access.json
Copying /clients
Deleted extra entry /clients/client1.json (purge is on)
Deleted extra entry /clients/client2.json (purge is on)
Copying /containers
Copying /cookbook_artifacts
Copying /cookbooks
Copying /data_bags
Copying /environments
Copying /invitations.json
Copying /members.json
Copying /nodes
Deleted extra entry /nodes/node2.json (purge is on)
Deleted extra entry /nodes/node1.json (purge is on)
Copying /org.json
Copying /policies
Copying /policy_groups
Copying /roles
Copying /acls/organization.json
Copying /acls/clients/4thcafe-validator.json
Copying /acls/containers/clients.json
Copying /acls/containers/containers.json
Copying /acls/containers/cookbook_artifacts.json
Copying /acls/containers/cookbooks.json
Copying /acls/containers/data.json
Copying /acls/containers/environments.json
Copying /acls/containers/groups.json
Copying /acls/containers/nodes.json
Copying /acls/containers/policies.json
Copying /acls/containers/policy_groups.json
Copying /acls/containers/roles.json
Copying /acls/containers/sandboxes.json
Copying /acls/environments/_default.json
Copying /acls/groups/000000000000e6f11cb43723b8ba5c35.json
Copying /acls/groups/admins.json
Copying /acls/groups/clients.json
Copying /acls/groups/users.json
Copying /groups/000000000000e6f11cb43723b8ba5c35.json
Copying /groups/admins.json
Copying /groups/clients.json
Updated /groups/clients.json
Copying /groups/users.json
** Finished **
root@api:~# ls /backup/organizations/4thcafe/acls/nodes/
node1.json  node2.json
root@api:~# ls /backup/organizations/4thcafe/acls/clients/
4thcafe-validator.json  client1.json  client2.json
root@api:~# ls /backup/user_acls
janedoe.json  pivotal.json
root@api:~# mv /backup/ /backup1
```

lbakerchef commented:
Test of fix, `ec backup`, patched code, chef-server 12.19.31.
The fix appears to work.
```
# vagrant box / 12.19.31 / ubuntu1804 / post-patch / test of fix, ec backup

root@api:~# chef-server-ctl version
12.19.31
root@api:~# knife node list
node1
node2
root@api:~# knife client list
4thcafe-validator
client1
client2
root@api:~# chef-server-ctl user-list
janedoe
pivotal
user1
user2
root@api:~# knife ec backup /backup2 --purge -V -s https://api.chef-server.dev
root@api:~# knife node delete node1 -y
Deleted node[node1]
root@api:~# knife node delete node2 -y
Deleted node[node2]
root@api:~# knife client delete client1 -y
Deleted client[client1]
root@api:~# knife client delete client2 -y
Deleted client[client2]
root@api:~# chef-server-ctl user-delete user1 -y
Checking organization memberships...
Deleting user user1.
root@api:~# chef-server-ctl user-delete user2 -y
Checking organization memberships...
Deleting user user2.
root@api:~# knife node list

root@api:~# knife client list
4thcafe-validator
root@api:~# chef-server-ctl user-list
janedoe
pivotal
root@api:~# ls /backup2/organizations/4thcafe/acls/clients/
4thcafe-validator.json  client1.json  client2.json
root@api:~# ls /backup2/organizations/4thcafe/acls/nodes/
node1.json  node2.json
root@api:~# ls /backup2/user_acls
janedoe.json  pivotal.json  user1.json  user2.json
root@api:~# knife ec backup /backup2 --purge -V -s https://api.chef-server.dev
Downloading Users
Deleting user user1 from local backup (purge is on)
Deleting user user2 from local backup (purge is on)
Downloading organization object for 4thcafe from https://api.chef-server.dev/organizations/4thcafe
Copying /acls/groups/billing-admins.json
Copying /groups/billing-admins.json
Copying /groups/admins.json
Copying /acls/groups/public_key_read_access.json
Copying /groups/public_key_read_access.json
Copying /acls
Deleted extra entry /acls/nodes/node1.json (purge is on)
Deleted extra entry /acls/nodes/node2.json (purge is on)
Deleted extra entry /acls/clients/client1.json (purge is on)
Deleted extra entry /acls/clients/client2.json (purge is on)
Copying /clients
Deleted extra entry /clients/client1.json (purge is on)
Deleted extra entry /clients/client2.json (purge is on)
Copying /containers
Copying /cookbook_artifacts
Copying /cookbooks
Copying /data_bags
Copying /environments
Copying /groups
Updated /groups/clients.json
Copying /invitations.json
Copying /members.json
Copying /nodes
Deleted extra entry /nodes/node2.json (purge is on)
Deleted extra entry /nodes/node1.json (purge is on)
Copying /org.json
Copying /policies
Copying /policy_groups
Copying /roles
** Finished **
root@api:~# ls /backup2/organizations/4thcafe/acls/nodes/
root@api:~# ls /backup2/organizations/4thcafe/acls/clients/
4thcafe-validator.json
root@api:~# ls /backup2/user_acls
janedoe.json  pivotal.json
root@api:~# diff --brief --recursive /backup1 /backup2
Only in /backup1/organizations/4thcafe/acls/clients: client1.json
Only in /backup1/organizations/4thcafe/acls/clients: client2.json
Only in /backup2/organizations/4thcafe/acls: cookbook_artifacts
Only in /backup2/organizations/4thcafe/acls: cookbooks
Only in /backup2/organizations/4thcafe/acls: data_bags
Only in /backup1/organizations/4thcafe/acls/nodes: node1.json
Only in /backup1/organizations/4thcafe/acls/nodes: node2.json
Only in /backup2/organizations/4thcafe/acls: policies
Only in /backup2/organizations/4thcafe/acls: policy_groups
Only in /backup2/organizations/4thcafe/acls: roles
```

lbakerchef commented:
Test of `ec restore`, chef-server 12.19.31
Success unknown
```
# vagrant box / 12.19.31 / ubuntu1804 / post-patch / restore test

root@api:~# knife node list

root@api:~# knife client list
4thcafe-validator
root@api:~# chef-server-ctl user-list
janedoe
pivotal
root@api:~# knife node create node1
Created node[node1]
root@api:~# knife node create node2
Created node[node2]
root@api:~# knife client create client1
Created client[client1]
root@api:~# knife client create client2
Created client[client2]
root@api:~# chef-server-ctl user-create user1 user 1 user1@example.com 'abc123'
root@api:~# chef-server-ctl user-create user2 user 2 user2@example.com 'abc123'
root@api:~# knife node list
node1
node2
root@api:~# knife client list
4thcafe-validator
client1
client2
root@api:~# chef-server-ctl user-list
janedoe
pivotal
user1
user2
root@api:~# knife ec backup /backup --purge -V -s https://api.chef-server.dev
root@api:~# mv /backup /vagrant/dotfiles/backup
root@api:~# exit
logout
vagrant@api:~$ exit
logout
Connection to 127.0.0.1 closed.
mac:dev lbaker$ VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant destroy && VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant up chef-server
vagrant@api:~$ sudo -i
root@api:~# dpkg -i /vagrant/dotfiles/chef-server-core_12.19.31-1_amd64.deb
Selecting previously unselected package chef-server-core.
(Reading database ... 39785 files and directories currently installed.)
Preparing to unpack .../chef-server-core_12.19.31-1_amd64.deb ...
Unpacking chef-server-core (12.19.31-1) ...
Setting up chef-server-core (12.19.31-1) ...
root@api:~# chef-server-ctl reconfigure -y 
Chef Client finished, 495/1084 resources updated in 03 minutes 45 seconds
Chef Server Reconfigured!
root@api:~# chef-server-ctl version
12.19.31
root@api:~# alias knife='/opt/opscode/bin/knife'
root@api:~# export EDITOR=vim
root@api:~# mkdir .chef
root@api:~# chef-server-ctl user-create janedoe Jane Doe janed@example.com 'abc123' --filename ~/.chef/janedoe.pem
root@api:~# chef-server-ctl org-create 4thcafe 'Fourth Cafe, Inc.' --association_user janedoe --filename ~/.chef/4thcafe-validator.pem
root@api:~# echo "[default]
> client_name     = 'janedoe'
> client_key      = '~/.chef/janedoe.pem'
> chef_server_url = 'https://api.chef-server.dev/organizations/4thcafe'">~/.chef/credentials
root@api:~# knife ssl fetch
WARNING: Certificates from api.chef-server.dev will be fetched and placed in your trusted_cert
directory (/root/.chef/trusted_certs).

Knife has no means to verify these are the correct certificates. You should
verify the authenticity of these certificates after downloading.

Adding certificate for api_chef-server_dev in /root/.chef/trusted_certs/api_chef-server_dev.crt
root@api:~# knife ssl check
Connecting to host api.chef-server.dev:443
Successfully verified certificates from `api.chef-server.dev'
root@api:~# knife node list

root@api:~# knife client list
4thcafe-validator
root@api:~# chef-server-ctl user-list
janedoe
pivotal
root@api:~# knife ec restore /vagrant/dotfiles/backup/
Restoring users
WARNING: Skipping pivotal user.  To overwrite pivotal, pass --overwrite-pivotal.
Restoring organization[4thcafe]
Restoring org admin data
Copying /groups/admins.json
Copying /groups/billing-admins.json
Copying /groups/public_key_read_access.json
Copying /acls/groups/billing-admins.json
Copying /acls/groups/public_key_read_access.json
Restoring the rest of the org
Copying /clients
Updated /clients/4thcafe-validator.json
Created /clients/client2.json
Created /clients/client1.json
Copying /containers
Copying /cookbook_artifacts
Copying /cookbooks
Copying /data_bags
Copying /environments
Copying /invitations.json
Copying /members.json
Copying /nodes
Created /nodes/node2.json
Created /nodes/node1.json
Copying /org.json
Updated /org.json
Copying /policies
Copying /policy_groups
Copying /roles
Copying /groups/000000000000e6f11cb43723b8ba5c35.json
Created /groups/000000000000e6f11cb43723b8ba5c35.json
Copying /groups/admins.json
Copying /groups/clients.json
Copying /groups/users.json
Updated /groups/users.json
Copying /groups/000000000000e6f11cb43723b8ba5c35.json
Updated /groups/000000000000e6f11cb43723b8ba5c35.json
Copying /groups/admins.json
Copying /groups/clients.json
Copying /groups/users.json
Copying /acls/groups/000000000000e6f11cb43723b8ba5c35.json
Copying /acls/groups/admins.json
Copying /acls/groups/clients.json
Copying /acls/groups/users.json
Copying /acls/clients
Updated /acls/clients/client1.json
Updated /acls/clients/client2.json
Updated /acls/clients/client2.json
Updated /acls/clients/client1.json
Copying /acls/containers
Copying /acls/cookbook_artifacts
Copying /acls/cookbooks
Copying /acls/data_bags
Copying /acls/environments
Copying /acls/nodes
Updated /acls/nodes/node2.json
Updated /acls/nodes/node2.json
Updated /acls/nodes/node1.json
Updated /acls/nodes/node1.json
Copying /acls/organization.json
Copying /acls/policies
Copying /acls/policy_groups
Copying /acls/roles
Copying /groups/admins.json
Copying /groups/billing-admins.json
Copying /groups/public_key_read_access.json
Restoring user ACLs
WARNING: Skipping pivotal user.  To overwrite pivotal, pass --overwrite-pivotal.
** Finished **
root@api:~# knife node list
ERROR: Failed to authenticate to https://api.chef-server.dev/organizations/4thcafe as janedoe with key /root/.chef/janedoe.pem
Response:  Invalid signature for user or client 'janedoe'
root@api:~# knife ssl fetch
WARNING: Certificates from api.chef-server.dev will be fetched and placed in your trusted_cert
directory (/root/.chef/trusted_certs).

Knife has no means to verify these are the correct certificates. You should
verify the authenticity of these certificates after downloading.

Adding certificate for api_chef-server_dev in /root/.chef/trusted_certs/api_chef-server_dev.crt
root@api:~# knife ssl check
Connecting to host api.chef-server.dev:443
Successfully verified certificates from `api.chef-server.dev'
root@api:~# knife node list
ERROR: Failed to authenticate to https://api.chef-server.dev/organizations/4thcafe as janedoe with key /root/.chef/janedoe.pem
Response:  Invalid signature for user or client 'janedoe'
```

lbakerchef commented:
Test of ec restore, chef-server 14.11.36
Success unknown
```
# vagrant box / 14.11.36 / ubuntu1804 / post-patch / backup-restore test

mac:dev lbaker$ VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant destroy && VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant up chef-server
vagrant@api:~$ sudo -i
root@api:~# dpkg -i /vagrant/dotfiles/chef-server-core_14.11.36-1_amd64.deb
Selecting previously unselected package chef-server-core.
(Reading database ... 39785 files and directories currently installed.)
Preparing to unpack .../chef-server-core_14.11.36-1_amd64.deb ...
Unpacking chef-server-core (14.11.36-1) ...
Setting up chef-server-core (14.11.36-1) ...
Thank you for installing Chef Infra Server!
root@api:~# chef-server-ctl reconfigure -y 
Chef Client finished, 495/1084 resources updated in 03 minutes 45 seconds
Chef Server Reconfigured!
root@api:~# chef-server-ctl version
14.11.36
root@api:~# alias knife='/opt/opscode/bin/knife'
root@api:~# export EDITOR=vim
root@api:~# mkdir .chef
root@api:~# chef-server-ctl user-create janedoe Jane Doe janed@example.com 'abc123' --filename ~/.chef/janedoe.pem
root@api:~# chef-server-ctl org-create 4thcafe 'Fourth Cafe, Inc.' --association_user janedoe --filename ~/.chef/4thcafe-validator.pem
root@api:~# echo "[default]
> client_name     = 'janedoe'
> client_key      = '~/.chef/janedoe.pem'
> chef_server_url = 'https://api.chef-server.dev/organizations/4thcafe'">~/.chef/credentials
root@api:~# knife ssl fetch
WARNING: Certificates from api.chef-server.dev will be fetched and placed in your trusted_cert
directory (/root/.chef/trusted_certs).

Knife has no means to verify these are the correct certificates. You should
verify the authenticity of these certificates after downloading.

Adding certificate for api_chef-server_dev in /root/.chef/trusted_certs/api_chef-server_dev.crt
root@api:~# knife ssl check
Connecting to host api.chef-server.dev:443
Successfully verified certificates from `api.chef-server.dev'
root@api:~# knife node create node1
Created node[node1]
root@api:~# knife node create node2
Created node[node2]
root@api:~# knife client create client1
Created client[client1]
root@api:~# knife client create client2
Created client[client2]
root@api:~# chef-server-ctl user-create user1 user 1 user1@example.com 'abc123'
root@api:~# chef-server-ctl user-create user2 user 2 user2@example.com 'abc123'
root@api:~# knife node list
node1
node2
root@api:~# knife client list
4thcafe-validator
client1
client2
root@api:~# chef-server-ctl user-list
janedoe
pivotal
user1
user2
root@api:~# knife ec backup /vagrant/dotfiles/backup14x --purge -V -s https://api.chef-server.dev
root@api:~# exit
logout
vagrant@api:~$ exit
logout
Connection to 127.0.0.1 closed.
mac:dev lbaker$ VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant destroy && VAGRANT_VAGRANTFILE=Vagrantfile.praj vagrant up chef-server
mac:dev lbaker$ vagrant ssh
vagrant@api:~$ sudo -i
root@api:~# dpkg -i /vagrant/dotfiles/chef-server-core_14.11.36-1_amd64.deb
root@api:~# chef-server-ctl reconfigure -y 
Chef Server Reconfigured!
root@api:~# chef-server-ctl version
14.11.36
root@api:~# alias knife='/opt/opscode/bin/knife'
root@api:~# export EDITOR=vim
root@api:~# mkdir .chef
root@api:~# chef-server-ctl user-create janedoe Jane Doe janed@example.com 'abc123' --filename ~/.chef/janedoe.pem
root@api:~# chef-server-ctl org-create 4thcafe 'Fourth Cafe, Inc.' --association_user janedoe --filename ~/.chef/4thcafe-validator.pem
root@api:~# echo "[default]
> client_name     = 'janedoe'
> client_key      = '~/.chef/janedoe.pem'
> chef_server_url = 'https://api.chef-server.dev/organizations/4thcafe'">~/.chef/credentials
root@api:~# knife ssl fetch
WARNING: Certificates from api.chef-server.dev will be fetched and placed in your trusted_cert
directory (/root/.chef/trusted_certs).

Knife has no means to verify these are the correct certificates. You should
verify the authenticity of these certificates after downloading.

Adding certificate for api_chef-server_dev in /root/.chef/trusted_certs/api_chef-server_dev.crt
root@api:~# knife ssl check
Connecting to host api.chef-server.dev:443
Successfully verified certificates from `api.chef-server.dev'
root@api:~# knife node list

root@api:~# knife client list
4thcafe-validator
root@api:~# chef-server-ctl user-list
janedoe
pivotal
root@api:~# knife ec restore /vagrant/dotfiles/backup14x/
Restoring users
WARNING: Skipping pivotal user.  To overwrite pivotal, pass --overwrite-pivotal.
Restoring organization[4thcafe]
Restoring org admin data
Copying /groups/admins.json
Copying /groups/billing-admins.json
Copying /groups/public_key_read_access.json
Copying /acls/groups/billing-admins.json
Copying /acls/groups/public_key_read_access.json
Restoring the rest of the org
Copying /clients
Updated /clients/4thcafe-validator.json
Created /clients/client2.json
Created /clients/client1.json
Copying /containers
Copying /cookbook_artifacts
Copying /cookbooks
Copying /data_bags
Copying /environments
Copying /invitations.json
Copying /members.json
Copying /nodes
Created /nodes/node1.json
Created /nodes/node2.json
Copying /org.json
Updated /org.json
Copying /policies
Copying /policy_groups
Copying /roles
Copying /groups/000000000000161dd515d1af1e371c96.json
Created /groups/000000000000161dd515d1af1e371c96.json
Copying /groups/admins.json
Copying /groups/clients.json
Copying /groups/users.json
Updated /groups/users.json
Copying /groups/000000000000161dd515d1af1e371c96.json
Updated /groups/000000000000161dd515d1af1e371c96.json
Copying /groups/admins.json
Copying /groups/clients.json
Copying /groups/users.json
Copying /acls/groups/000000000000161dd515d1af1e371c96.json
Copying /acls/groups/admins.json
Copying /acls/groups/clients.json
Copying /acls/groups/users.json
Copying /acls/clients
Updated /acls/clients/client1.json
Updated /acls/clients/client2.json
Updated /acls/clients/client2.json
Updated /acls/clients/client1.json
Copying /acls/containers
Copying /acls/environments
Copying /acls/nodes
Updated /acls/nodes/node1.json
Updated /acls/nodes/node2.json
Updated /acls/nodes/node2.json
Updated /acls/nodes/node1.json
Copying /acls/organization.json
Copying /groups/admins.json
Copying /groups/billing-admins.json
Copying /groups/public_key_read_access.json
Restoring user ACLs
WARNING: Skipping pivotal user.  To overwrite pivotal, pass --overwrite-pivotal.
** Finished **
root@api:~# knife node list
ERROR: Failed to authenticate to https://api.chef-server.dev/organizations/4thcafe as janedoe with key /root/.chef/janedoe.pem
Response:  Invalid signature for user or client 'janedoe'
```

PrajaktaPurohit reviewed (approved):
Thank you!
